### DISCUSSION TITLE
Vitest to be able to run native node:test tests

### DISCUSSION THREAD

**Discussion by ThePlenkov** [vitest-dev/vitest#6482] [Ideas]
Hi! I have a following idea

I would like to write test in framework independent way just using node:test package, I'm able to run those tests with `node --test` command.

Now I want to see how my tests behave with vitest. So what I expect:
- `npx vitest`
- vitest imports my files, but at the same time it resolves node:test as it would be vitest:test
- The whole scripts gets execute as it would be vitest.

What do you think of this proposal?

**Comment by ThePlenkov** [comment]
I've created some PoC https://github.com/ThePlenkov/node-test-with-anything/pull/1

the idea is following

- I run command like `node --import vitest/import.mjs --test` 
where import.mjs is 
```
import { register } from 'node:module';
register('./hooks.mjs', import.meta.url);
```

and  hooks.mjs is
```
export async function resolve(specifier, context, nextResolve) {
    // Take an `import` or `require` specifier and resolve it to a URL.    
    return nextResolve(specifier === "node:test" ? 'vitest' : specifier);
}
```

So idea is that it shoud execute vitest instead of node:test

However currently I have a following error:
```
Debugger attached.
Waiting for the debugger to disconnect...
file:///workspaces/node-test-with-anything/node_modules/vitest/dist/chunks/utils.Ck2hJTRs.js:5
    throw new Error(errorMsg);
          ^

Error: Vitest failed to access its internal state.

One of the following is possible:
- "vitest" is imported directly without running "vitest" command
- "vitest" is imported inside "globalSetup" (to fix this, use "setupFiles" instead, because "globalSetup" runs in a different context)
- Otherwise, it might be a Vitest bug. Please report it to https://github.com/vitest-dev/vitest/issues

    at getWorkerState (file:///workspaces/node-test-with-anything/node_modules/vitest/dist/chunks/utils.Ck2hJTRs.js:5:11)
    at getCurrentEnvironment (file:///workspaces/node-test-with-anything/node_modules/vitest/dist/chunks/utils.Ck2hJTRs.js:19:17)
    at createExpect (file:///workspaces/node-test-with-anything/node_modules/vitest/dist/chunks/vi.D6IHiKAI.js:475:20)
    at file:///workspaces/node-test-with-anything/node_modules/vitest/dist/chunks/vi.D6IHiKAI.js:520:22
    at ModuleJob.run (node:internal/modules/esm/module_job:262:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:482:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:117:5)

```

**Comment by sheremet-va** [comment]
Vitest is not a simple script you can just replace your other test runner with. Vitest requires its own module context to transform modules with Vite and store the global state.

`node:test` has a different interface and runs tests differently. It doesn't make sense for us to develop and support a version of the runner that doesn't use our own runner.

If you for some reason just want to replace one module with another (redirect `node:test` to `vitest`), you can use aliases: https://vitest.dev/config/#alias

**Comment by ThePlenkov** [comment]
@sheremet-va Thank you for a quick reply Vladimir. Actually I have tried it works 

```
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    alias:{
        "node:test":"vitest"    }
  },
})
```

I understand that vitest api is different from node:test API, but may be as an idea for vitest team to think of something like 

```
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    alias:{
        "node:test":"vitest:node"    }
  },
})
```

where vitest:node is a node:test compatible vitest API. Theroretically must not be that much work to have, but in this case you'll be able to onboard quickly devs from other frameworks which are using native node api.

**Reply by ThePlenkov** [reply]
So along with this 
![image](https://github.com/user-attachments/assets/b3e315a5-f5c2-4ff0-b56c-e33e02a7562e)
it may come with something like Node native compatible

**Reply by sheremet-va** [reply]
This is not something we are looking to implement.

**Reply by ThePlenkov** [reply]
I see. Meanwhile I found another testing library which is working in the way I was searching for:
https://github.com/tapjs/node-test-example

**Reply by ThePlenkov** [reply]
it is called "Interoperability with the Node.js Built-in Test Runner"
https://node-tap.org/node-test-interop/

Just keeping here this word to make it searchable in the project by word "Interoperability ". May be there will be more people asking for similar thing
